### PR TITLE
Split up pins.h to make pins more manageable

### PR DESCRIPTION
**Pull request by thinkyhead**
Hopefully this is helpful to organization and future expansion….

### CODE REVIEW

**Comment by CONSULitAS** [comment]
@MagoKimbra I Think this would be great. For languages_xxx.h, too. But have a look at https://github.com/sudar/Arduino-Makefile/issues/249 and just google for <arduino include subfolder>.

It does not work now with Arduino. It would break support for the Arduino IDE.

There are discussions for years, but the collaborators of the Arduino project don't seem to see the point. :-(

If you have a way to realize it, please tell.

**Comment by CONSULitAS** [comment]
@thinkyhead Thanks for the great "pins_K8200.h". I think it's the right way, because you can see how easy it is to add a new board if it's identical to an other.

And "pins_3Drag" shows how easy it is to add a board that is similar to an other.

**Comment by MagoKimbra** [comment]
@CONSULitAS Yes i see... Ok!!!

**Comment by alexborro** [comment]
It is nice, but all "pins_" files should be inside a folder "pins_mapping"... And the language files as well (may be "translation" folder).. The root folder is becoming a mess, too many files..

**Comment by bjmckenz** [comment]
+1 — however, the dir should be “config” and the files should be named “system.h” (e.g. “ramps_13_EFB.h”, “ultimaker2.h”).

I recommend “config” and not “pins” because I can see this evolving a bit. Can you imagine putting more things than pins there? I can. Given some of these are actual machines and some are motherboards, I see the machine people moving things like bed dimensions into the config files.

In other words, there are hardware features and software options. I’d like to group the feature-related things together, and keep the options (what a person tweaks) to a minimum so that they are easy to distinguish.

Admittedly the downside of this is that if a new feature or #define is added it must be spread out to all the config files. 

Just a thought…

**Comment by alexborro** [comment]
We can have different folders for different things.. Like "Pins_Mapping" for pins_*.h, "Machine_Specifics" for software configs related to specific machines and so one.. splitting things in folders are a good idea.

**Comment by thinkyhead** [comment]
@alexborro @bjmckenz Subfolders are not supported by Arduino's compiler, or we'd be using them already.

**Comment by alexborro** [comment]
Damn it.. What age do these arduino's guys live?
Em 01/02/2015 02:33, "Scott Lahteine" notifications@github.com escreveu:

**Comment by CONSULitAS** [comment]
@alexborro To answer your question, have a look at this:

arduino/Arduino#1004
sudar/Arduino-Makefile#251
sudar/Arduino-Makefile#249

:-)))

**Comment by thinkyhead** [comment]
@alexborro @MagoKimbra Ok, so yes I discovered we can move things into subfolders. But, we have to "pretend" they are libraries. Here's the trick.
1. Make a **libraries** folder inside the **Marlin** folder.
2. Make a **MarlinPins** folder inside **Marlin/libraries**.
3. Move all the **pins** files into this folder.
4. Rename **pins.h** to **MarlinPins.h**.
5. Add `#include <MarlinPins.h>` to **Marlin.ino** and **Marlin.pde**.
6. Replace `#include "pins.h"` to `#include <MarlinPins.h>` everywhere else.

I just did this with the **pins** and **language** files ("MarlinLanguage"), and it worked great.

It also needs some lines in the Makefile for command-line build....

```
VPATH += ./libraries/MarlinPins
VPATH += ./libraries/MarlinLanguage
```

**Comment by thinkyhead** [comment]
#1537

**Comment by alexborro** [comment]
I'm ok with such solution.
It is transparent to end user and make things more organized.
So far we can have:
Libraries/MarlinPins
Libraries/MarlinLanguages

And I think we can apply the same idea for thermistor tables as well.

Cheers.

Alex.

2015-02-24 4:18 GMT-03:00 Scott Lahteine notifications@github.com:

> #1537 https://github.com/MarlinFirmware/Marlin/pull/1537
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/MarlinFirmware/Marlin/pull/1416#issuecomment-75710240
> .

## 

_"Não é o mais forte da espécie que sobrevive, nem o mais inteligente. É
aquele que se adapta melhor as mudanças"  ( Charles Darwin )_

Alex Borro

**Comment by MagoKimbra** [comment]
I tried to download the marlin with the libraries folder, but when i compile do it error because it can not find the libraries MarlinLanguage.h. I have Arduino IDE 1.6.0

**Comment by thinkyhead** [comment]
@MagoKimbra Hmm, I'm not too surprised. I closed my PR for the time-being because I got stuck on building with the Makefile. But it was working for me in the IDE, both 1.5.2 and 1.0.5. I will download 1.6 and try it later!

**Comment by thinkyhead** [comment]
I have a new update of `temperature.cpp` (etc.) pending with #1541 and that might behave differently.
